flink-ml-lib/src/main/java/org/apache/flink/ml/feature/hashingtf/HashingTF.java

```diff
@@ -0,0 +1,193 @@
+ * default. Users could also enforce all non-zero output values as 1 by setting {@link
+ * HashingTFParams#BINARY} as true.
+ *
+ * <p>For the hashing trick, see https://en.wikipedia.org/wiki/Feature_hashing for details.
+ */
+public class HashingTF implements Transformer<HashingTF>, HashingTFParams<HashingTF> {
+    private final Map<Param<?>, Object> paramMap = new HashMap<>();
+
+    private static final org.apache.flink.shaded.guava30.com.google.common.hash.HashFunction
+            HASH_FUNC = murmur3_32(0);
+
+    public HashingTF() {
+        ParamUtils.initializeMapWithDefaultValues(paramMap, this);
+    }
+
+    @Override
+    public Table[] transform(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+
+        ResolvedSchema tableSchema = inputs[0].getResolvedSchema();
+        RowTypeInfo inputTypeInfo = TableUtils.getRowTypeInfo(tableSchema);
+        RowTypeInfo outputTypeInfo =
+                new RowTypeInfo(
+                        ArrayUtils.addAll(inputTypeInfo.getFieldTypes(), VectorTypeInfo.INSTANCE),
+                        ArrayUtils.addAll(inputTypeInfo.getFieldNames(), getOutputCol()));
+
+        DataStream<Row> output =
+                tEnv.toDataStream(inputs[0])
+                        .map(
+                                new HashTFFunction(getInputCol(), getBinary(), getNumFeatures()),
+                                outputTypeInfo);
+        return new Table[] {tEnv.fromDataStream(output)};
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        ReadWriteUtils.saveMetadata(this, path);
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return paramMap;
+    }
+
+    public static HashingTF load(StreamTableEnvironment tEnv, String path) throws IOException {
+        return ReadWriteUtils.loadStageParam(path);
+    }
+
+    /** The main logic of {@link HashingTF}, which converts the input to a sparse vector. */
+    public static class HashTFFunction implements MapFunction<Row, Row> {
+        private final String inputCol;
+        private final boolean binary;
+        private final int numFeatures;
+
+        public HashTFFunction(String inputCol, boolean binary, int numFeatures) {
+            this.inputCol = inputCol;
+            this.binary = binary;
+            this.numFeatures = numFeatures;
+        }
+
+        @Override
+        public Row map(Row row) throws Exception {
+            Object inputObj = row.getField(inputCol);
+
+            Iterable<Object> inputList;
+            if (inputObj.getClass().isArray()) {
+                inputList = Arrays.asList((Object[]) inputObj);
+            } else if (inputObj instanceof Iterable) {
+                inputList = (Iterable<Object>) inputObj;
+            } else {
+                throw new IllegalArgumentException(
```

Comment:
It might be better to do this check in `transform()` method and throw exceptions as early as during compilation.

Reply:
Unfortunately we cannot do this check during compilation. If the input is a list, the typeinformation turns to be `generciType(ArrayList)`, which is hard to throw exceptions during compilation process.